### PR TITLE
Implement Smrnmi extension

### PR DESCRIPTION
**Pull request by aswaterman**
We don't model any sources of RNMI, so this is mostly vestigial.

cc @ptomsich

### CODE REVIEW

**Comment by aswaterman** [comment]
The spec for this feature is a chapter in the privileged ISA manual, BTW.  https://github.com/riscv/riscv-isa-manual/blob/faa77da9701329bc212241d5f9c04684358bb28f/src/rnmi.tex

File: riscv/csrs.h [hunk 1]
```diff
@@ -256,6 +256,20 @@ class mstatus_csr_t final: public base_status_csr_t {
 
 typedef std::shared_ptr<mstatus_csr_t> mstatus_csr_t_p;
 
+class mnstatus_csr_t final: public csr_t {
```
Comment (scottj97) [inline riscv/csrs.h]: This could be `basic_csr_t` and only override `unlogged_write()`. That would be more consistent with the other simple CSRs.

File: riscv/insns/mnret.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+require_extension(EXT_SMRNMI);
+require_privilege(PRV_M);
+set_pc_and_serialize(p->get_state()->mnepc->read());
+reg_t s = STATE.mnstatus->read();
+reg_t prev_prv = get_field(s, MNSTATUS_MNPP);
+reg_t prev_virt = get_field(s, MNSTATUS_MNPV);
+if (prev_prv != PRV_M) {
+  reg_t mstatus = STATE.mstatus->read();
+  mstatus = set_field(mstatus, MSTATUS_MPRV, 0);
+  p->put_csr(CSR_MSTATUS, mstatus);
```
Comment (scottj97) [inline riscv/insns/mnret.h]: Why not `STATE.mstatus->write(mstatus)`? That would avoid an unnecessary map lookup in `csrmap`. (Same comment for the other `put_csr()` just below.)

Could even add a simple method to `mstatus_csr_t` so this could do `STATE.mstatus->clear_mprv()`.

File: riscv/mmu.h [hunk 3]
```diff
@@ -406,6 +406,7 @@ class mmu_t
   inline bool in_mprv()
   {
     return proc != nullptr
+           && !(proc->state.mnstatus && !get_field(proc->state.mnstatus->read(), MNSTATUS_NMIE))
```
Comment (scottj97) [inline riscv/mmu.h:409]: Would it make sense/improve anything if we always instantiated `state.mnstatus` and then only put it into `csrmap` if Smrnmi is enabled. Then we wouldn't need to test for null in at least three places I see.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/mmu.h (hunk 3 above)
Comment (aswaterman) [reply riscv/mmu.h:409]: The only reason I didn't do that to begin with is that the reset value of the NMIE bit is 0 if the extension is implemented, but to suppress the behavior we need to set it to 1.  I'll try it and see how I feel about it.

File: riscv/csrs.h (hunk 1 above)
Comment (aswaterman) [reply riscv/csrs.h]: Thanks for the suggestion.  I had forgotten about `basic_csr_t`.

File: riscv/insns/mnret.h (hunk 2 above)
Comment (aswaterman) [reply riscv/insns/mnret.h]: Copy-pasta from the implementation of the MRET instruction.

I decided against defining `clear_mprv`, since inlining the code isn't so bad and I don't expect the method to get much reuse.

File: riscv/mmu.h (hunk 3 above)
Comment (aswaterman) [reply riscv/mmu.h:409]: I don't think it turns out to be an improvement.  Since we need NMIE to be 1 to suppress the new behavior, we'll have a conundrum if we ever add devices to Spike that can generate NMIs: NMIs would appear to be enabled even though the extension isn't present, so we'd need a special-case check over there instead.

File: riscv/csrs.cc [hunk 4]
```diff
@@ -516,6 +516,24 @@ reg_t mstatus_csr_t::compute_mstatus_initial_value() const noexcept {
...
 
+// implement class mnstatus_csr_t
+mnstatus_csr_t::mnstatus_csr_t(processor_t* const proc, const reg_t addr):
+  basic_csr_t(proc, addr, 0) {
+}
+
+bool mnstatus_csr_t::unlogged_write(const reg_t val) noexcept {
+  // NMIE can be set but not cleared
+  const reg_t mask = (~read() & MNSTATUS_NMIE)
+                   | (proc->extension_enabled('H') ? MNSTATUS_MNPV : 0)
```
Comment (scottj97) [inline riscv/csrs.cc:527]: I haven't read the spec but I'm guessing this MNPV should be 0 if no hypervisor? What if we write it to 1, and then disable H?

Most bits that are H-only are masked on read, not write, for this reason. Or written to 0 when H is cleared in `misa_csr_t::unlogged_write()`.

File: riscv/csrs.h [hunk 5]
```diff
@@ -256,6 +256,15 @@ class mstatus_csr_t final: public base_status_csr_t {
 
 typedef std::shared_ptr<mstatus_csr_t> mstatus_csr_t_p;
 
+class mnstatus_csr_t final: public basic_csr_t {
+ public:
+  mnstatus_csr_t(processor_t* const proc, const reg_t addr);
+ protected:
+  virtual bool unlogged_write(const reg_t val) noexcept override;
+ private:
+  reg_t val;
```
Comment (scottj97) [inline riscv/csrs.h]: Hides the parent class's `val`. Instead call `basic_csr_t::unlogged_write()` from within `mnstatus_csr_t::unlogged_write()`.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/csrs.h (hunk 5 above)
Comment (aswaterman) [reply riscv/csrs.h]: Oops, I should've deleted this when I changed the choice of base class.

File: riscv/csrs.cc (hunk 4 above)
Comment (aswaterman) [reply riscv/csrs.cc:527]: I'm following our implementation of `mstatus.MPV`.  Is the implication that needs to change, too?

Comment (aswaterman) [reply riscv/csrs.cc:527]: Nevermind, I'm not accurately following our implementation of `mstatus.MPV`; I missed that writing `misa` also alters that bit.  I've updated the patch to mimic that behavior.

**Review by scottj97 — APPROVED** [review]

**Comment by aswaterman** [comment]
@scottj97 Thanks for helping me improve this PR.
